Add ros rviz files for icub

Yeshasvitvs opened the pull request:
Given that yarp supports ros, it is beneficial to make use of  [ros rviz](http://wiki.ros.org/rviz) visualization tool to visualize icub. This will be helpful in doing demos with the real robot along with the rviz visualization on big screens. This PR adds the necessary ros launch file and rviz configuration files for visualizing icub models. 

This addressed https://github.com/dic-iit/lab-events-demos/issues/102

@traversaro

traversaro commented:
>  This addressed dic-iit/lab-events-demos#102

@Yeshasvitvs not all the users of the icub-models open source repository have access to your private repositories, so it would be great if you could discuss the rationale behind this PR in the PR itself, thanks!

traversaro commented on iCub/ros/rviz/iCubRviz.rviz:
```diff
@@ -0,0 +1,733 @@
+    collapsed: false
+  Tool Properties:
+    collapsed: false
+  Width: 2560
```
This is quite big for a  regular resolution screen. Can we fit  it in something more normal, like ~1000 px?

Yeshasvitvs replied:
The resolution is because I was testing on a big monitor.
I will change it to a lower resolution screen.

Yeshasvitvs replied:
Done

traversaro commented on iCub/ros/rviz/iCubRviz.rviz:
```diff
@@ -0,0 +1,733 @@
+  Height: 1388
+  Hide Left Dock: false
+  Hide Right Dock: false
+  QMainWindow State: 000000ff00000000fd00000004000000000000016a000004b4fc0200000008fb0000001200530065006c0065006300740069006f006e00000001e10000009b0000007601000003fb0000001e0054006f006f006c002000500072006f007000650072007400690065007302000001ed000001df00000185000000a3fb000000120056006900650077007300200054006f006f02000001df000002110000018500000122fb000000200054006f006f006c002000500072006f0070006500 ...
```
This variable is quite obscure and is not clear its role (it would be quite difficult to understand its changes in PRs). Can we just omit it so that RViz uses some kind of "default" vaue?

Yeshasvitvs replied:
Again this is a variable from autogeneration of this configuration file. 
I am not certain if we can omit this variable. I will test it and update here.

Yeshasvitvs replied:
Deleting that variable opens rviz in a different configuration. 

![Screenshot from 2019-06-04 09-52-17](https://user-images.githubusercontent.com/6505998/58861457-b4f69080-86ae-11e9-8673-b354572940b6.png)

traversaro replied:
Ok, it was worth giving it a try.

traversaro commented on iCub/ros/rviz/iCubRviz.rviz:
```diff
@@ -0,0 +1,733 @@
+          Alpha: 1
+          Show Axes: false
+          Show Trail: false
+          Value: true
```
Not all the icub models have  all those links. Are all this options  compulsory? Can we just omit them?

Yeshasvitvs replied:
This is again the result of autogenrated `.rviz` configuration file as there is a robot model added inside rviz. I tested it with `iCubGenova02` and `iCubGazeboV2_5` urdf models.

traversaro replied:
Yes, but what happens if somebody uses it with the `iCubGenova01` that for example does not contain the `l_forearm_skin_7` URDF "link"? What happens if in the future we add or remove links from the model? I would strongly prefer to have a conf file that we do not need to update for each change in the iCub model, and if it is not possible we should document how this file can be regenerated (if it cannot be regenerated automatically). 
However, according to http://wiki.ros.org/rviz/DisplayTypes/RobotModel this option seems to be optional, so I think it should be possible to just remove them without any particular problem (also simplifying sensibly the conf file).

Yeshasvitvs replied:
> However, according to http://wiki.ros.org/rviz/DisplayTypes/RobotModel this option seems to be optional, so I think it should be possible to just remove them without any particular problem (also simplifying sensibly the conf file).

Yes this option is not strictly necessary but the idea of having a ros package with a launch file and rviz configuration is that some one without much experience with ros or rivz can simple launch it without much hassle to add the robot model and then add the proper [prefix](https://github.com/Yeshasvitvs/icub-models/blob/7347e37aa715c27e8eecbaa90b82e5c42ca41c82/iCub/ros/rviz/iCubRviz.rviz#L664), which is set in `yarprobostatepublisher` parameters. I believe in this case it is better to leave a model in the rviz configurations as at least the users who will be using the rviz visualization tool actively are mostly (at least for the time being) users of latest models (DIC members).

Yeshasvitvs replied:
Is it enough to put description and instructions in a readme file so that other users can modify it if required? One of the main goals of this PR is the simplification of launching demos. Having the model already there makes it faster and smoother.

traversaro replied:
Sorry, I did not want to remove the model. I just wanted to remove the optional per-link configuration, such as: 
~~~
        base_link:
          Alpha: 1
          Show Axes: false
          Show Trail: false
~~~
that can create problems if a different model is loaded. Do you think this can create problems in any way?

traversaro commented on iCub/ros/rviz/iCubRviz.rviz:
```diff
@@ -0,0 +1,733 @@
+    Tree Height: 1043
+  - Class: rviz/Selection
+    Name: Selection
+  - Class: rviz/Tool Properties
```
Is this panel required?

Yeshasvitvs replied:
The `.rviz` file is generate from rviz by saving the configuration. So, all the contents of the file are generated automatically and I did not manually edit the file.

Yeshasvitvs replied:
Deleting the panels sections opens rviz without any panels by default. Users can go to the panel drop down menu and add but I think it is unnecessary. So, I prefer to leave the panels section as such.

![Screenshot from 2019-06-04 09-55-49](https://user-images.githubusercontent.com/6505998/58861681-39e1aa00-86af-11e9-9e2d-8445936a23ca.png)

Yeshasvitvs replied:
`- Class: rviz/Tool Properties` section can be omitted safely. Rviz starts normally (i.e. without any changes in the panels) probably by loading default values for tools panel. Removed in https://github.com/robotology/icub-models/pull/23/commits/82b1b41444a8bdd1f29cd41fd25edce00fd23b62

Yeshasvitvs commented:
> > This addressed [dic-iit/lab-events-demos#102](https://github.com/dic-iit/lab-events-demos/issues/102)
> 
> @Yeshasvitvs not all the users of the icub-models open source repository have access to your private repositories, so it would be great if you could discuss the rationale behind this PR in the PR itself, thanks!

@traversaro I hoped the comment "This PR adds ros rviz launch file and configuration files for visualizing icub models." is supposed to convey the rationale behind this PR. I will update the description to be a bit more elaborate.

traversaro commented:
> @traversaro I hoped the comment "This PR adds ros rviz launch file and configuration files for visualizing icub models." is supposed to convey the rationale behind this PR. I will update the description to be a bit more elaborate.

Yes, but for netiquette I would avoid to reference private repos from public repos. You can always link pubic repos issues from private repos, and thanks to GitHub's features the reference will also automatically appear in the public repo, at least for people with access to the specific private repo.

traversaro commented:
I think we should document in the README how this files are supposed to be used. Ideally we should also refer to the ROS distribution with which we tested those files.

traversaro commented:
@fiorisi What do you think about this PR? Given the generality of this file, I think it make sense to keep it in this repo, even if it is not an autogenerated file (in a sense, it is similar to the `model.config.in` file ( https://github.com/robotology/icub-models/blob/master/model.config.in ). 
However, all the files contained in `iCub` folder up to know have been autogenerated (see https://github.com/robotology/icub-model-generator/blob/master/.travis.yml#L153), and placing non-autogenerated files in the same directory of automatic generated files may be confusing. Given that we do not support anymore using directly the models from the source repository, perhaps it could make sense to leave the `iCub` directory to just contain autogenerated files, and have a different directory for non-autogenerated files, with some specific CMake rules for installing them (similarly to what is happening now with the `model.config.in` one).

fiorisi commented:
@traversaro I agree, we can keep these configuration files in this repo. As you suggest probably we can create a folder for non-autogenerated files.

traversaro commented:
> @traversaro I agree, we can keep these configuration files in this repo. As you suggest probably we can create a folder for non-autogenerated files.
In this case we only need to add a line similar to this one to copy the folder?

Yes.

fjandrad commented:
What is missing to merge? @Yeshasvitvs @traversaro

fiorisi commented:
If I am not mistaken the [readme](https://github.com/robotology/icub-models/pull/23#issuecomment-498721951) and the [CMake file](https://github.com/robotology/icub-models/pull/23#issuecomment-499368393) shall be updated.
Furthermore the files shall be updated as described in [this comment](https://github.com/robotology/icub-models/pull/23#issuecomment-498723962).

Yeshasvitvs commented:
Readme added https://github.com/robotology/icub-models/pull/23/commits/3f3ed4be09749410c071095d355cc52f90458f2e

Yeshasvitvs commented:
The files for rviz visualization are put out of the auto generated `iCub` directory and the `CMakeLists.txt` is updated to install them correctly inside the `iCub` at the installation destination. 
https://github.com/robotology/icub-models/pull/23/commits/2a66592ee1167c74c50e3222e94907a97ae13977

Yeshasvitvs commented:
Currently we run tranform sever from the [configuration file](https://github.com/robotology/human-dynamics-estimation/blob/master/conf/xml/TransformServer.xml) inside HDE. However, to make the ros configuration file standalone in this repo, it is better to have a transform server configuration file with in this repo. Added the configuration file and updated the readme with all the steps needed to run rviz for icub visualization https://github.com/robotology/human-dynamics-estimation/blob/master/conf/xml/TransformServer.xml

Yeshasvitvs commented:
@traversaro I believe all necessary changes are made for the PR. Please review it and let me know if there are any more changes to be done.

Yeshasvitvs commented:
@traversaro @fiorisi One more thing I noticed is that the auto generated files doesn't include any configuration file for ros wrapper (check [this file](https://github.com/Yeshasvitvs/icub-models/blob/feature/iCubGazeboV2_5_fixed/iCub/conf/gazebo_icub_ros.ini) for an example). I believe it is better to include within icub-models repo as a lot of testing is done in simulation first, having the option to work with simulation models and rviz is useful.

Yeshasvitvs commented:
Ping @traversaro

traversaro commented:
> @traversaro I believe all necessary changes are made for the PR. Please review it and let me know if there are any more changes to be done.

As mentioned by @fiorisi in https://github.com/robotology/icub-models/pull/23#issuecomment-512684886 , the comment https://github.com/robotology/icub-models/pull/23#issuecomment-498723962 needs to be addressed.

traversaro commented:
Furthermore, as already discussed in https://github.com/robotology/icub-models/pull/23#discussion_r293272721, can you check if removing all those defaults for all the frames (i.e. to be clear, remove all the lines from https://github.com/robotology/icub-models/blob/05d749d696fd529bc2d31e0f7e7419ad6162bce5/ros/rviz/iCubRviz.rviz#L56 to https://github.com/robotology/icub-models/blob/05d749d696fd529bc2d31e0f7e7419ad6162bce5/ros/rviz/iCubRviz.rviz#L650) still result in a usable model? If it is not possible to have a configuration that works for all iCub models without manual modifications for each model, then I think it is better to not include the conf file at all, or it is better to explicit write for which models is compatible.

Yeshasvitvs commented:
> > @traversaro I believe all necessary changes are made for the PR. Please review it and let me know if there are any more changes to be done.
> 
> As mentioned by @fiorisi in [#23 (comment)](https://github.com/robotology/icub-models/pull/23#issuecomment-512684886) , the comment [#23 (comment)](https://github.com/robotology/icub-models/pull/23#issuecomment-498723962) needs to be addressed.

@traversaro I am not sure which comments you mean exactly. 

- The config files are put out of the autogenerated directory as mentioned in https://github.com/robotology/icub-models/pull/23#issuecomment-517209909 and the corresponding commit is https://github.com/robotology/icub-models/commit/2a66592ee1167c74c50e3222e94907a97ae13977
- Readme is updated in https://github.com/robotology/icub-models/pull/23/commits/0eb9e04f5f0db9f883cee21e62dbad0890ef1579#diff-af632284272fce7902b28d153c8e7bd2

Yeshasvitvs commented:
> Furthermore, as already discussed in [#23 (comment)](https://github.com/robotology/icub-models/pull/23#discussion_r293272721), can you check if removing all those defaults for all the frames (i.e. to be clear, remove all the lines from
> 
> https://github.com/robotology/icub-models/blob/05d749d696fd529bc2d31e0f7e7419ad6162bce5/ros/rviz/iCubRviz.rviz#L56
> 
> to
> https://github.com/robotology/icub-models/blob/05d749d696fd529bc2d31e0f7e7419ad6162bce5/ros/rviz/iCubRviz.rviz#L650
> 
> ) still result in a usable model? If it is not possible to have a configuration that works for all iCub models without manual modifications for each model, then I think it is better to not include the conf file at all, or it is better to explicit write for which models is compatible.

@traversaro the link names are removed from the rviz configuration file in the commit https://github.com/robotology/icub-models/pull/23/commits/acabbcdbfcfa56742faf8c02e02c3c8d926d2c38

Rviz is launched without any errors and I believe the link names are taken according to the urdf model passed as an [argument in the ros launch file](https://github.com/Yeshasvitvs/icub-models/blob/acabbcdbfcfa56742faf8c02e02c3c8d926d2c38/ros/launch/iCubRviz.launch#L7). So there wouldn't be any problems in launching different robot models.

traversaro commented:
Great, thanks!
